### ISSUE TITLE
How to get depth form disparity map

### ISSUE THREAD

**Issue by ChuangbinC** [mrharicot/monodepth#118]
Hi, I want test the precision of pretrained model which use KITTI data. For example I take this picture into the model. **How do I convert the disparity map to depth?** 

**the input picture:**
![image](https://user-images.githubusercontent.com/18183832/36845068-3e4cc93c-1d90-11e8-8fee-91c851883d18.png)

**Comment by mrharicot** [comment]
Hi,
I would not expect a model trained on KITTI to produce anything sensible on NYUv2.

To obtain depth, you need to convert disparity using the following formula:
`depth = baseline * focal / disparity`
For KITTI the baseline is 0.54m and the focal ~721 pixels. The relative disparity outputted by the model has to be scaled by 1242 which is the original image size.
The final formula is:
`depth = 0.54 * 721 / (1242 * disp)`

However it is unclear how to convert this disparity into depth for images that do not match KITTI's focal and aspect ratio.
The bottom line is, first invert the disparity. If the result doesn't look awful you can try to find a scale by hand to match this to depth in meters.

**Comment by CCB307** [comment]
@mrharicot Thank for your answer! 
1. As you mean, if I build a small data set by myself and **is it no sensible to put this data set into the model trained on KITTI?**
2. If I want to test the precision of this network, I have to use my data set to train a new model and evaluate it?

**Comment by mrharicot** [comment]
I am not sure what you mean by this. You can fine tune a model trained on KITTI with your own data. The model will do better if the new dataset looks like KITTI though.

**Comment by CCB307** [comment]
@mrharicot Thank you! I will try your method.

**Comment by CCB307** [comment]
@mrharicot Hi,I am so sorry for troubling you again. 
As you say, I need to convert disparity to depth using following formula:
`depth = baseline * focal / disparity`
My model is trained on KITTI and then I collect some dataset looks like KITTI using my own camera.
What I want to know is that **the focal and baseline in the final formula are same as KITTI or they are related to my camera.**

**Comment by keishatsai** [comment]
Hi @CCB307, I think that I can answer this question to you.

Focal length and baseline are vary due to the different cameras, so you have to find the specific one.
As far as I know, there are no  focal length and baseline involve in the model architecture, I suggest that you can try the other dataset for inference which the focal length and baseline are already known, like Cityscapes. 
You'll know what to do :)

**Comment by ChuangbinC** [comment]
@keishatsai Thank you every much! 
Maybe I am wrong because I think that I can use this model to estimate depth from a single image before.But the monocular camera has no baseline.

My project is that estimating depth from single image.

**Comment by mrharicot** [comment]
Hi,
In theory you should use the focal and baselines of the dataset you trained on, in this case either KITTI or Cityscapes.
However when your test data is quite different from the training data (focal or aspect ratio) the network might not output anything reasonable anyway, and it becomes less clear what is reasonable to do.

**Comment by au1khan** [comment]
Can we say that, our depth is as accurate as the prior information about the environment of the testing image we have? I mean focal and baseline.

**Comment by mrharicot** [comment]
You can always convert the disparity to a depth by inverting it.
If the network produces something reasonable then you can assume it is correct up to scale.

**Comment by ChuangbinC** [comment]
@mrharicot I am sorry to trouble you again. I want to know **why the output disparity need to be multiplied by 1242 which is the original image size ?**
If I use the output disparity map (size is 256x512) to compute the depth map, the formula is that :
```
depth = baseline * focal / disparity
```
or :
```
depth = baseline * focal / (512*disparity)
```

**Comment by mrharicot** [comment]
You need to scale by the _original_ image size because the baseline is with respect to the _original_ image size.

**Comment by ChuangbinC** [comment]
As you mean, if I want to get the depth map, I need to resize the disparity map（256x512）to the original image size at first ?

**Comment by keishatsai** [comment]
@ChuangbinC, yes, you have to. The disparity is normalised by its width, so you have to scale it back in order to get the real depth instead of relative depth.

**Comment by Ariel-JUAN** [comment]
@keishatsai 
Hi, the max disparity is 0.3, so when I convert to depth, the min depth = 0.54 * 721 / (1242 * disp) = 1.04
So what does it mean? I am confused.
I want to generate a depth map like this:
![2](https://user-images.githubusercontent.com/30546375/39994688-989481cc-57ac-11e8-9030-b1f33b225afe.png)
But all I got is this:
![3](https://user-images.githubusercontent.com/30546375/39994721-ace19b06-57ac-11e8-821d-b9650c03cfa1.png)
I wonder how I can generate the right depth map.

**Comment by mrharicot** [comment]
Hi,
You need to set the vmax of the colormap to 80m when showing it.
Most methods tend to show disparity/inverse depth for kitti, as they are easier to read.

**Comment by Ariel-JUAN** [comment]
@mrharicot 
Thank you for the reply.
I test a image, and get a disparity map. Then I resize the disparity to (1242,375).
Next I want to generate the depth,   depth = 0.54 * 721 / (1242 * disp) 
The code is like this 
    disp_pp = post_process_disparity(disp.squeeze()).astype(np.float32)
    disp_pp = cv2.resize(disp_pp, (original_width, original_height))
    depth = 0.54*721/(1242*disp_pp)
I get the depth value is about 4.13~498.222. So what does it mean?
I don't know how to match the max depth=80m
Thanks

**Comment by mrharicot** [comment]
Hi,
When visualising using matplotlib you can specify a maximum value:
plt.imshow(depth, vmax=80)

**Comment by Ariel-JUAN** [comment]
@mrharicot Thanks, the depth value is about 4.13~498.222. What does it mean? The real depth?

**Comment by Ariel-JUAN** [comment]
@mrharicot  Thanks!

**Comment by bellakate** [comment]
Sir @mrharicot Thank You for your commendable work
My Question is, How do we know the depth of a particular object e.g car in an image from the depth map?

**Comment by bellakate** [comment]
@Ariel-JUAN I am getting the same distorted depth map as your's. Can  you please suggest any solution?

**Comment by kingsleyluoxin** [comment]
> Hi,
> I would not expect a model trained on KITTI to produce anything sensible on NYUv2.
> 
> To obtain depth, you need to convert disparity using the following formula:
> `depth = baseline * focal / disparity`
> For KITTI the baseline is 0.54m and the focal ~721 pixels. The relative disparity outputted by the model has to be scaled by 1242 which is the original image size.
> The final formula is:
> `depth = 0.54 * 721 / (1242 * disp)`
> 
> However it is unclear how to convert this disparity into depth for images that do not match KITTI's focal and aspect ratio.
> The bottom line is, first invert the disparity. If the result doesn't look awful you can try to find a scale by hand to match this to depth in meters.

Hello, for 1242 * disp , what is the size of disp?

**Comment by hemangchawla** [comment]
> Hi,
> I would not expect a model trained on KITTI to produce anything sensible on NYUv2.
> 
> To obtain depth, you need to convert disparity using the following formula:
> `depth = baseline * focal / disparity`
> For KITTI the baseline is 0.54m and the focal ~721 pixels. The relative disparity outputted by the model has to be scaled by 1242 which is the original image size.
> The final formula is:
> `depth = 0.54 * 721 / (1242 * disp)`
> 
> However it is unclear how to convert this disparity into depth for images that do not match KITTI's focal and aspect ratio.
> The bottom line is, first invert the disparity. If the result doesn't look awful you can try to find a scale by hand to match this to depth in meters.

How would this work with MonoDepth2?

**Comment by mrharicot** [comment]
monodepth2 predicts depth directly.
It needs to be scaled to obtain metric depth however.

**Comment by sven97** [comment]
Thank you for your wonderful work!
May I ask how can I obtain metric depth map using monodepth2?

**Comment by mrharicot** [comment]
Have a look at:
https://github.com/nianticlabs/monodepth2/blob/master/evaluate_depth.py

**Comment by qureshinomaan** [comment]
> Hi,
> I would not expect a model trained on KITTI to produce anything sensible on NYUv2.
> 
> To obtain depth, you need to convert disparity using the following formula:
> `depth = baseline * focal / disparity`
> For KITTI the baseline is 0.54m and the focal ~721 pixels. The relative disparity outputted by the model has to be scaled by 1242 which is the original image size.
> The final formula is:
> `depth = 0.54 * 721 / (1242 * disp)`
> 
> However it is unclear how to convert this disparity into depth for images that do not match KITTI's focal and aspect ratio.
> The bottom line is, first invert the disparity. If the result doesn't look awful you can try to find a scale by hand to match this to depth in meters.

If I have the disparity value in a grayscale matrix, is it the same as pixels?

**Comment by yokosyun** [comment]
I have idea to confirm it

focal = 7.070912e+02
baseline = 0.54
image_scale = 512

depth = focal * baseline / (disp *image_scale)

print("torch.mean(depth)=",torch.mean(depth))
print("torch.max(depth)=",torch.max(depth))
print("torch.min(depth)=",torch.min(depth))



and compare with your maxdisparity(=192)

depth = 7.070912e+02 * 0.54 / 192
          = 1.988694

check whether it is close to your min depth and the value never be less than 1.988694. if not change your image_scale = 1 or 512 or 1242

**Comment by rafiqul713** [comment]
Thanks all. This information is really helpful. What are the baseline and focal of the NYU v2 dataset?

**Comment by sainivedh** [comment]
For monodepth2 you get direct depths from disp_to_depth function. Just need to multiply by scaling factor to get metric depth
